packages/react-ui-core/src/Gmap/FreeDrawLayer.js

```diff
@@ -0,0 +1,218 @@
+      shape.push(
+        poly.getPath()
+          .getAt(i)
+          .toUrlValue(5)
+          .split(',')
+          .map(Number)
+          .reverse()
+      )
+    }
+    return shape
+  }
+
+  enableMapControls() {
+    const { map, mapControls } = this.props
+    map.setOptions({
+      draggable: true,
+      zoomControl: true,
+      scrollwheel: true,
+      disableDoubleClickZoom: true,
+      ...mapControls,
+    })
+    this.setState({
+      drawing: false,
+    })
+  }
+
+  disableMapControls() {
+    const { map } = this.props
+    map.setOptions({
+      draggable: false,
+      zoomControl: false,
+      scrollwheel: false,
+      disableDoubleClickZoom: false,
+    })
+    this.setState({
+      drawing: true,
+    })
+  }
+
+  @autobind
+  enableMapDraw() {
+    const { map, multipleShapes } = this.props
+    this.disableMapControls()
+
+    if (!multipleShapes) {
+      this.clearAllShapes()
+    }
+    setupEvents(map, EVENTS, {
+      onMouseDown: () => {
+        this.drawFreeHand()
+      },
+    }, true)
+  }
+
+  drawFreeHand() {
+    const { map, onMapDrawStart, dataStyle } = this.props
+    const poly = new window.google.maps.Polyline({
+      map,
+      clickable: false,
+      ...dataStyle.polyline,
+    })
+    const drawPoly = new window.google.maps.Polygon({
+      map,
+      path: poly.getPath(),
+      clickable: false,
+      ...dataStyle.polylineFill,
+    })
+
+    if (onMapDrawStart) {
+      onMapDrawStart()
+    }
+
+    const listener = setupEvents(map, EVENTS, {
+      onMouseMove: e => {
+        poly.getPath().push(e.latLng)
+      },
+      onTouchStart: e => {
+        poly.getPath().push(e.latLng)
+      },
+    }, false)
+
+    const onMouseUp = () => {
+      removeEvent(listener.onMouseMove)
+      poly.setMap(null)
+      drawPoly.setMap(null)
+      this.addShape(poly.getPath())
+      this.enableMapControls()
+    }
+
+    setupEvents(map, EVENTS, {
+      onMouseUp,
+      onTouchEnd: onMouseUp,
+    }, true)
+  }
+
+  addShapeToGMaps(path) {
+    const { map, dataStyle } = this.props
+    return new window.google.maps.Polygon({
+      map,
+      path,
+      ...dataStyle.polygon,
+    })
+  }
+
+  addShape(path) {
+    const { onMapDrawEnd } = this.props
+    const { polygons } = this.state
+
+    const rawPolygon = this.addShapeToGMaps(path)
+
+    this.setState({
+      polygons: [...polygons, rawPolygon],
+    }, () => {
+      if (onMapDrawEnd) {
+        onMapDrawEnd(this.state.polygons.reduce((acc, polygon, i) => {
+          acc[i] = this.getPolygonCoords(polygon)
+          return acc
+        }, {}))
+      }
+    })
+  }
+
+  @autobind
+  clearAllShapes() {
+    const { polygons } = this.state
+
+    for (let i = 0; i < polygons.length; i += 1) {
+      polygons[i].setMap(null)
+    }
+    this.setState({
+      polygons: {},
+    })
+  }
+
+  formatLongLatToGmapsCoordinates(shapes) {
+    return Object.values(shapes)
+      .reduce((polygonAcc, polygon) => {
```

Comment:
This seems complicated to reason about - we are reducing inside a reducer - can we simplify the data we are storing  or just break this up into smaller functions that are explicit in what they are doing?

Reply:
Ill look into this

Reply:
We discussed with Orion and he proposed some really nice functional things that we'd want to discuss with Tad before attempting to include it in react-ui. For now we have decided to leave the reduce within the reduce. Thanks @yasabere and @orionstein for the call - learned a lot about ramda and FP 👍